xdsl/dialects/complex.py

```diff
@@ -1,41 +1,105 @@
 from __future__ import annotations
 
 import abc
-from typing import ClassVar, cast
+from collections.abc import Sequence
+from typing import ClassVar, Generic, cast
+
+from typing_extensions import TypeVar
 
 from xdsl.dialects.arith import FastMathFlagsAttr
 from xdsl.dialects.builtin import (
-    AnyAttr,
     AnyFloat,
     AnyFloatConstr,
     ArrayAttr,
     ArrayOfConstraint,
     ComplexType,
     FloatAttr,
+    FloatData,
     IntegerAttr,
     IntegerType,
-    ParamAttrConstraint,
 )
 from xdsl.interfaces import ConstantLikeInterface
-from xdsl.ir import Attribute, Dialect, Operation, SSAValue
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    Operation,
+    ParametrizedAttribute,
+    SSAValue,
+)
 from xdsl.irdl import (
+    AnyAttr,
     AnyOf,
     EqIntConstraint,
     IRDLOperation,
+    ParamAttrConstraint,
     RangeOf,
     VarConstraint,
     base,
+    irdl_attr_definition,
     irdl_op_definition,
     operand_def,
+    param_def,
     prop_def,
     result_def,
     traits_def,
 )
+from xdsl.parser import AttrParser
+from xdsl.printer import Printer
 from xdsl.traits import Pure
 from xdsl.utils.exceptions import VerifyException
+from xdsl.utils.hints import isa
 
 ComplexTypeConstr = ComplexType.constr(AnyFloat)
 
+_ComplexNumberElementType = TypeVar(
+    "_ComplexNumberElementType",
+    bound=AnyFloat,
+    covariant=True,
+    default=AnyFloat,
+)
+
+
+@irdl_attr_definition
+class ComplexNumberAttr(ParametrizedAttribute, Generic[_ComplexNumberElementType]):
+    name = "complex.number"
+
+    real: FloatData = param_def(converter=FloatData.get)
+    imag: FloatData = param_def(converter=FloatData.get)
+    type: ComplexType[_ComplexNumberElementType]
+
+    def print_parameters(self, printer: Printer) -> None:
+        with printer.in_angle_brackets():
+            printer.print_string(
+                f":{self.type.element_type} {self.real.data}, {self.imag.data}"
+            )
+        printer.print_string(f" : {self.type}")
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[Attribute]:
+        """
+        Example:
+        ```mlir
+        #complex.number<:f64 1.0, 2.0> : complex<f64>
+        ```
+        """
+        with parser.in_angle_brackets():
+            parser.parse_punctuation(":")
+            assert isa(element_type := parser.parse_type(), AnyFloat)
+            real = FloatData(parser.parse_float())
+            parser.parse_punctuation(",")
+            imag = FloatData(parser.parse_float())
+        parser.parse_punctuation(":")
+        assert isa(
+            (complex_type := parser.parse_type()),
+            ComplexType[AnyFloat],
+        )
+        assert complex_type.element_type == element_type
```

Comment:
Asserting and ValueError are for programmer error, we should avoid raising them for incorrect user input. Could you please change these to raise ParseError instead?